index.js

```diff
@@ -1,80 +1,106 @@
-		var issuerContext = this._module && this._module.issuer && this._module.issuer.context || context;
-		var relativeUrl = issuerContext && path.relative(issuerContext, filePath).split(path.sep).join("/");
-		var relativePath = relativeUrl && path.dirname(relativeUrl) + "/";
-		if (~relativePath.indexOf("../")) {
-			outputPath = path.posix.join(outputPath, relativePath, url);
-		} else {
-			outputPath = relativePath + url;
-		}
-		url = relativePath + url;
-	} else if (config.outputPath) {
-		// support functions as outputPath to generate them dynamically
-		outputPath = (
-			typeof config.outputPath === "function"
-			? config.outputPath(url)
-			: config.outputPath + url
-		);
-		url = outputPath;
-	} else {
-		outputPath = url;
-	}
+  if (config.useRelativePath) {
+    // Only the dirname is needed in this case.
+    outputPath = outputPath.replace(url, "");
 
-	var publicPath = "__webpack_public_path__ + " + JSON.stringify(url);
-	if (config.publicPath) {
-		// support functions as publicPath to generate them dynamically
-		publicPath = JSON.stringify(
-			typeof config.publicPath === "function"
-			? config.publicPath(url)
-			: config.publicPath + url
-		);
-	}
+    // We have access only to entry point relationships. So we work with this relations.
+    var issuerContext = (this._module && this._module.issuer && this._module.issuer.context) || context;
+    var relativeUrl = issuerContext && path.relative(issuerContext, this.resourcePath);
+    var relativePath = relativeUrl && path.dirname(relativeUrl);
+    var outputDirname = relativePath.replace(/\.\.(\/|\\)/g, "").split(path.sep).join("/");
 
-	if (query.emitFile === undefined || query.emitFile) {
-		this.emitFile(outputPath, content);
-	}
+    // Output path
+    // If the `outputDirname` is pointing to up in relation to the `outputPath`.
+    // We forced him to the webpack output path config. Even though it is empty.
+    if (outputDirname.indexOf(outputPath) !== 0) {
+      outputDirname = outputPath;
+    }
+    outputPath = path.join(outputDirname, url).split(path.sep).join("/");
 
-	return "module.exports = " + publicPath + ";";
+    // Public path
+    // Entry files doesn't pass through the `file-loader`.
+    // So we don't have access to the files context to compare with your assets context
+    // then we need to kidnap from `extract-text-webpack-plugin` if you using him and the same way
+    // force the `relativePath` to extract files on the webpack output path config folder.
+    // otherwise make the same using the same output path from text extracted
+    // with `extract-text-webpack-plugin`.
+    var output = this.options.output || {};
+    if (output.filename && output.filename !== "extract-text-webpack-plugin-output-filename") {
+      relativePath = outputDirname;
+    } else if (toString.call(config.textOutputPath) === "[object String]") {
+      var outputPackageDirname = output.path.replace(this.options.context + path.sep, "");
```

Comment:
See the comment directly above, please group `issuer, relative, output` via an `{Object}`